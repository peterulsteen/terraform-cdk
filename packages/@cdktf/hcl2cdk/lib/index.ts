import { parse } from "@cdktf/hcl2json";
import generate from "@babel/generator";
import template from "@babel/template";
import * as t from "@babel/types";
import prettier from "prettier";
import { pascalCase, camelCase } from "change-case";
import { schema, Output, Variable, Provider, Module } from "./schema";
import { DirectedGraph } from "graphology";
import * as rosetta from "jsii-rosetta";
import {
  Reference,
  extractReferencesFromExpression,
  referencesToAst,
} from "./expressions";

const valueToTs = (item: any): t.Expression => {
  switch (typeof item) {
    case "string":
      return referencesToAst(item, extractReferencesFromExpression(item));
    case "boolean":
      return t.booleanLiteral(item);
    case "number":
      return t.numericLiteral(item);
    case "object":
      if (Array.isArray(item)) {
        return t.arrayExpression(item.map(valueToTs));
      }

      if (
        Object.keys(item).includes("for_each") ||
        Object.keys(item).includes("count")
      ) {
        throw new Error(
          `Unsupported Terraform feature found: for-each loops are not yet supported: ${JSON.stringify(
            item
          )}`
        );
      }

      return t.objectExpression(
        Object.entries(item)
          .filter(([_key, value]) => value !== undefined)
          .map(([key, value]) =>
            t.objectProperty(t.stringLiteral(camelCase(key)), valueToTs(value))
          )
      );
  }
  throw new Error("Unsupported type " + item);
};

function findUsedReferences(
  nodeIds: string[],
  item: unknown,
  references: Reference[] = []
): Reference[] {
  if (Array.isArray(item)) {
    return [
      ...references,
      ...item.reduce(
        (carry, i) => [...carry, ...findUsedReferences(nodeIds, i)],
        []
      ),
    ];
  }

  if (typeof item === "object") {
    return [
      ...references,
      ...Object.values(item as Record<string, any>).reduce(
        (carry, i) => [...carry, ...findUsedReferences(nodeIds, i)],
        []
      ),
    ];
  }

  if (typeof item === "string") {
    const extractedRefs = extractReferencesFromExpression(item);
    if (!extractedRefs.every((ref) => nodeIds.includes(ref.referencee.id))) {
      throw new Error(
        `Found a reference that is unknown: ${item} resulted in references ${JSON.stringify(
          extractedRefs
        )}, but only these identifiers are known: ${JSON.stringify(nodeIds)}`
      );
    }

    return [...references, ...extractedRefs];
  }
  return references;
}

function asExpression(
  type: string,
  name: string,
  config: any,
  isReferenced: boolean
) {
  const isNamespacedImport = type.includes(".");
  const subject = isNamespacedImport
    ? t.memberExpression(
        t.identifier(type.split(".")[0]), // e.g. aws
        t.identifier(pascalCase(type.split(".")[1])) // e.g. NatGateway
      )
    : t.identifier(pascalCase(type));

  const expression = t.newExpression(subject, [
    t.thisExpression(),
    t.stringLiteral(name),
    valueToTs(config),
  ]);
  return isReferenced
    ? t.variableDeclaration("const", [
        t.variableDeclarator(t.identifier(camelCase(name)), expression),
      ])
    : t.expressionStatement(expression);
}
function output(key: string, _id: string, item: Output): t.Statement {
  const [{ value, description, sensitive }] = item;

  return asExpression(
    "TerraformOutput",
    key,
    {
      value,
      description,
      sensitive,
    },
    false
  );
}

function variable(
  key: string,
  id: string,
  item: Variable,
  graph: DirectedGraph
): t.Statement {
  // We don't handle type information right now
  const [{ type, ...props }] = item;

  return asExpression("TerraformVariable", key, props, isReferenced(graph, id));
}

function local(key: string, _id: string, item: any): t.Statement {
  return t.variableDeclaration("const", [
    t.variableDeclarator(t.identifier(camelCase(key)), valueToTs(item)),
  ]);
}

function modules(
  key: string,
  id: string,
  item: Module,
  graph: DirectedGraph
): t.Statement {
  const [{ source, ...props }] = item;
  return asExpression(source, key, props, isReferenced(graph, id));
}

function provider(key: string, _id: string, item: Provider): t.Statement {
  const props = item[0];

  if (props.alias) {
    throw new Error(
      `Unsupported Terraform feature found at "${key}": provider alias are not yet supported`
    );
  }

  return asExpression(
    `${key}.${pascalCase(key + "Provider")}`,
    key,
    props,
    false
  );
}
function isReferenced(graph: DirectedGraph, id: string) {
  return graph.outNeighbors(id).length > 0;
}

function resource(
  type: string,
  key: string,
  id: string,
  item: Provider,
  graph: DirectedGraph
): t.Statement {
  const [provider, ...name] = type.split("_");
  return asExpression(
    `${provider}.${name.join("_")}`,
    key,
    item[0],
    isReferenced(graph, id)
  );
}

// locals, provider, variables, and outputs are global key value maps
function forEachGlobal<T, R>(
  prefix: string,
  record: Record<string, T> | undefined,
  iterator: (key: string, id: string, value: T, graph: DirectedGraph) => R
): Record<string, (graph: DirectedGraph) => R> {
  return Object.entries(record || {}).reduce((carry, [key, item]) => {
    const id = `${prefix}.${key}`;
    return {
      ...carry,
      [id]: (graph: DirectedGraph) => iterator(key, id, item, graph),
    };
  }, {});
}

// data and resource are namespaced key value maps
function forEachNamespaced<T, R>(
  record: Record<string, Record<string, T>> | undefined,
  iterator: (
    type: string,
    key: string,
    id: string,
    value: T,
    graph: DirectedGraph
  ) => R,
  prefix?: string
): Record<string, (graph: DirectedGraph) => R> {
  return Object.entries(record || {}).reduce(
    (outerCarry, [type, items]) => ({
      ...outerCarry,
      ...Object.entries(items).reduce((innerCarry, [key, item]) => {
        const id = prefix ? `${prefix}.${type}.${key}` : `${type}.${key}`;
        return {
          ...innerCarry,
          [id]: (graph: DirectedGraph) => iterator(type, key, id, item, graph),
        };
      }, {} as Record<string, (graph: DirectedGraph) => R>),
    }),
    {} as Record<string, (graph: DirectedGraph) => R>
  );
}

export async function convertToTypescript(filename: string, hcl: string) {
  const json = await parse(filename, hcl);
  const plan = schema.parse(json);

  const nodeMap = {
    ...forEachGlobal("providers", plan.provider, provider),
    ...forEachGlobal("var", plan.variable, variable),
    // locals are a special case
    ...forEachGlobal(
      "local",
      Array.isArray(plan.locals)
        ? plan.locals.reduce((carry, locals) => ({ ...carry, ...locals }), {})
        : {},
      local
    ),
    ...forEachGlobal("out", plan.output, output),
    ...forEachGlobal("module", plan.module, modules),
    ...forEachNamespaced(plan.resource, resource),
    ...forEachNamespaced(plan.data, resource, "data"),
  };

  const graph = new DirectedGraph();
  // Add all nodes so we can detect if an edge is added for an unknown link
  Object.entries(nodeMap).forEach(([key, value]) =>
    graph.addNode(key, { code: value })
  );
  const nodeIds = Object.keys(nodeMap);

  // Add Edges
  function addGlobalEdges(_key: string, id: string, value: unknown) {
    findUsedReferences(nodeIds, value).forEach((ref) => {
      if (!graph.hasDirectedEdge(ref.referencee.id, id)) {
        graph.addDirectedEdge(ref.referencee.id, id);
      }
    });
  }
  function addNamespacedEdges(
    _type: string,
    _key: string,
    id: string,
    value: unknown
  ) {
    findUsedReferences(nodeIds, value).forEach((ref) => {
      if (!graph.hasDirectedEdge(ref.referencee.id, id)) {
        graph.addDirectedEdge(ref.referencee.id, id);
      }
    });
  }

  Object.values({
    ...forEachGlobal("providers", plan.provider, addGlobalEdges),
    ...forEachGlobal("var", plan.variable, addGlobalEdges),
    // locals are a special case
    ...forEachGlobal(
      "local",
      Array.isArray(plan.locals)
        ? plan.locals.reduce((carry, locals) => ({ ...carry, ...locals }), {})
        : {},
      addGlobalEdges
    ),
    ...forEachGlobal("out", plan.output, addGlobalEdges),
    ...forEachGlobal("module", plan.module, addGlobalEdges),
    ...forEachNamespaced(plan.resource, addNamespacedEdges),
    ...forEachNamespaced(plan.data, addNamespacedEdges, "data"),
  }).forEach((cb) => cb(graph));

  const expressions: t.Statement[] = [];
  let nodesToVisit = [...nodeIds];

  while (nodesToVisit.length > 0) {
    graph.forEachNode((nodeId) => {
      if (!nodesToVisit.includes(nodeId)) {
        return;
      }

      const unresolvedDependencies = graph
        .inNeighbors(nodeId)
        .filter((item) => nodesToVisit.includes(item));

      if (unresolvedDependencies.length === 0) {
        nodesToVisit = nodesToVisit.filter((id) => nodeId !== id);
        expressions.push(graph.getNodeAttribute(nodeId, "code")(graph));
      }
    });
  }

  const providerImports = Object.keys(plan.provider || {}).map(
    (providerName) =>
      template(
        `import * as ${providerName} from "./.gen/${providerName}"`
      )() as t.Statement
  );

  const moduleImports = Object.values(plan.module || {}).map(
    ([{ source }]) =>
      template(
        `import * as ${pascalCase(source)} from "./.gen/${source}"`
      )() as t.Statement
  );

  function gen(statements: t.Statement[]) {
    return prettier.format(generate(t.program(statements) as any).code, {
      parser: "babel",
    });
  }

  return {
    all: gen([...providerImports, ...moduleImports, ...expressions]),
    imports: gen([...providerImports, ...moduleImports]),
    code: gen(expressions),
  };
}

type File = { contents: string; fileName: string };
const translations = {
  typescript: (file: File) => file.contents,
  python: (file: File) =>
    rosetta.translateTypeScript(file, new rosetta.PythonVisitor()).translation,
  java: (file: File) =>
    rosetta.translateTypeScript(file, new rosetta.JavaVisitor()).translation,
  csharp: (file: File) =>
    rosetta.translateTypeScript(file, new rosetta.CSharpVisitor()).translation,
};

type ConvertOptions = {
  language: keyof typeof translations;
};

export async function convert(
  fileName: string,
  hcl: string,
  { language }: ConvertOptions
) {
  const translater = translations[language];

  if (!translater) {
    throw new Error("Unsupported language used: " + language);
  }
  const tsCode = await convertToTypescript(fileName, hcl);
  return {
    all: translater({ fileName, contents: tsCode.all }),
    imports: translater({ fileName, contents: tsCode.imports }),
    code: translater({ fileName, contents: tsCode.code }),
  };
}
