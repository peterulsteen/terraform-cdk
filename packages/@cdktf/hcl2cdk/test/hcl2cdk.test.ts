import { convert } from "../lib/index";

describe("convert", () => {
  describe("typescript", () => {
    it.each([
      [
        "output",
        `
        output "cidr_out" {
            value = "test"
        }`,
      ],
      [
        "sensitive output",
        `
        output "cidr_out" {
            value = "test"
            sensitive = true
        }`,
      ],
      [
        "output withdescription",
        `
        output "cidr_out" {
            value = "test"
            sensitive = true
            description = "Best output"
        }`,
      ],
      [
        "multiple outputs",
        `
        output "first_cidr_out" {
            value = "first"
        }
        output "second_cidr_out" {
            value = "second"
        }`,
      ],
      ["variable", `variable "weekday" {}`],
      [
        "variable with default",
        `
        variable "availability_zone_names" {
            type    = list(string)
            default = ["us-west-1a"]
            description = "What AMI to use to create an instance"
        }`,
      ],
      ["empty provider", `provider "docker" {}`],
      [
        "provider with complex config",
        `
        provider "aws" {
          access_key                  = "mock_access_key"
          region                      = "us-east-1"
          secret_key                  = "mock_secret_key"
          skip_credentials_validation = true
          skip_metadata_api_check     = true
          skip_requesting_account_id  = true
        
          endpoints {
            dynamodb = "http://localhost:8000"
          }
        }
      `,
      ],
      [
        "simple resource",
        `
        resource "aws_vpc" "example" {
          cidr_block = "10.0.0.0/16"
        }`,
      ],
      [
        "complex resource",
        ` 
        resource "aws_cloudfront_distribution" "s3_distribution" {
          origin {
            domain_name = "aws_s3_bucket.b.bucket_regional_domain_name"
            origin_id   = "local_s3_origin_id"
        
            s3_origin_config {
              origin_access_identity = "origin-access-identity/cloudfront/ABCDEFG1234567"
            }
          }
        
          enabled             = true
          is_ipv6_enabled     = true
          comment             = "Some comment"
          default_root_object = "index.html"
        
          logging_config {
            include_cookies = false
            bucket          = "mylogs.s3.amazonaws.com"
            prefix          = "myprefix"
          }
        
          aliases = ["mysite.example.com", "yoursite.example.com"]
        
          default_cache_behavior {
            allowed_methods  = ["DELETE", "GET", "HEAD", "OPTIONS", "PATCH", "POST", "PUT"]
            cached_methods   = ["GET", "HEAD"]
            target_origin_id = "local_s3_origin_id"
        
            forwarded_values {
              query_string = false
        
              cookies {
                forward = "none"
              }
            }
        
            viewer_protocol_policy = "allow-all"
            min_ttl                = 0
            default_ttl            = 3600
            max_ttl                = 86400
          }
        
          # Cache behavior with precedence 0
          ordered_cache_behavior {
            path_pattern     = "/content/immutable/*"
            allowed_methods  = ["GET", "HEAD", "OPTIONS"]
            cached_methods   = ["GET", "HEAD", "OPTIONS"]
            target_origin_id = "local_s3_origin_id"
        
            forwarded_values {
              query_string = false
              headers      = ["Origin"]
        
              cookies {
                forward = "none"
              }
            }
        
            min_ttl                = 0
            default_ttl            = 86400
            max_ttl                = 31536000
            compress               = true
            viewer_protocol_policy = "redirect-to-https"
          }
        
          # Cache behavior with precedence 1
          ordered_cache_behavior {
            path_pattern     = "/content/*"
            allowed_methods  = ["GET", "HEAD", "OPTIONS"]
            cached_methods   = ["GET", "HEAD"]
            target_origin_id = "local_s3_origin_id"
        
            forwarded_values {
              query_string = false
        
              cookies {
                forward = "none"
              }
            }
        
            min_ttl                = 0
            default_ttl            = 3600
            max_ttl                = 86400
            compress               = true
            viewer_protocol_policy = "redirect-to-https"
          }
        
          price_class = "PriceClass_200"
        
          restrictions {
            geo_restriction {
              restriction_type = "whitelist"
              locations        = ["US", "CA", "GB", "DE"]
            }
          }
        
          tags = {
            Environment = "production"
          }
        
          viewer_certificate {
            cloudfront_default_certificate = true
          }
        }`,
      ],
      [
        "simple data source",
        `
        data "aws_subnet" "selected" {
          id = "subnet_id"
        }`,
      ],
      [
        "locals",
        `
        locals {
          service_name = "forum"
          owner        = "Community Team"
          is_it_great  = true
          how_many     = 42
        }`,
      ],
      [
        "multiple locals blocks",
        `
        locals {
          service_name = "forum"
          owner        = "Community Team"
        }

        locals {
          is_it_great  = true
          how_many     = 42
        }`,
      ],
      [
        "resource references",
        `
        resource "aws_kms_key" "examplekms" {
          description             = "KMS key 1"
          deletion_window_in_days = 7
        }
        
        resource "aws_s3_bucket" "examplebucket" {
          bucket = "examplebuckettftest"
          acl    = "private"
        }
        
        resource "aws_s3_bucket_object" "examplebucket_object" {
          key        = "someobject"
          bucket     = aws_s3_bucket.examplebucket.id
          source     = "index.html"
          kms_key_id = aws_kms_key.examplekms.arn
        }`,
      ],
      [
        "locals references",
        `
        locals {
          bucket_name = "foo"
        }
        
        resource "aws_s3_bucket" "examplebucket" {
          bucket = local.bucket_name
          acl    = "private"
        }`,
      ],
      [
        "variable references",
        `
        variable "bucket_name" {
          type    = string
          default = "demo"
        }
        
        resource "aws_s3_bucket" "examplebucket" {
          bucket = var.bucket_name
          acl    = "private"
        }`,
      ],
      [
        "data references",
        `
        variable "bucket_name" {
          type    = string
          default = "demo"
        }
        
        data "aws_s3_bucket" "examplebucket" {
          bucket = var.bucket_name
          acl    = "private"
        }
        
        resource "aws_s3_bucket_object" "examplebucket_object" {
          key        = "someobject"
          bucket     = data.aws_s3_bucket.examplebucket.arn
          source     = "index.html"
        }`,
      ],
      [
        "double references",
        `
        variable "bucket_name" {
          type    = string
          default = "demo"
        }
        
        resource "aws_s3_bucket" "examplebucket" {
          bucket = var.bucket_name
          acl    = "private"
          tags = {
            tag-key = var.bucket_name
          }
        }`,
      ],
      [
        "modules",
        `
        module "vpc" {
          source = "terraform-aws-modules/vpc/aws"
        
          name = "my-vpc"
          cidr = "10.0.0.0/16"
        
          azs             = ["eu-west-1a", "eu-west-1b", "eu-west-1c"]
          private_subnets = ["10.0.1.0/24", "10.0.2.0/24", "10.0.3.0/24"]
          public_subnets  = ["10.0.101.0/24", "10.0.102.0/24", "10.0.103.0/24"]
        
          enable_nat_gateway = true
          enable_vpn_gateway = true
        
          tags = {
            Terraform = "true"
            Environment = "dev"
          }
        }`,
      ],
      [
        "referenced modules",
        `
        module "vpc" {
          source = "terraform-aws-modules/vpc/aws"
        
          name = "my-vpc"
          cidr = "10.0.0.0/16"
        
          azs             = ["eu-west-1a", "eu-west-1b", "eu-west-1c"]
          private_subnets = ["10.0.1.0/24", "10.0.2.0/24", "10.0.3.0/24"]
          public_subnets  = ["10.0.101.0/24", "10.0.102.0/24", "10.0.103.0/24"]
        
          enable_nat_gateway = true
          enable_vpn_gateway = true
        
          tags = {
            Terraform = "true"
            Environment = "dev"
          }
        }
        
        output "subnet_ids" {
          value = module.vpc.public_subnets
        }`,
      ],
    ])("%s configuration", async (_name, hcl) => {
      const { all } = await convert(`file.hcl`, hcl, {
        language: "typescript",
      });
      expect(all).toMatchSnapshot();
    });

    describe("errors on", () => {
      it.each([
        [
          "resource references with HCL functions",
          `
          resource "aws_kms_key" "examplekms" {
            description             = "KMS key 1"
            deletion_window_in_days = 7
          }
          
          resource "aws_s3_bucket" "examplebucket" {
            bucket = "examplebuckettftest"
            acl    = "private"
          }
          
          resource "aws_s3_bucket_object" "examplebucket_object" {
            key        = "someobject"
            bucket     = element(aws_s3_bucket.examplebucket, 0).id
            source     = "index.html"
            kms_key_id = aws_kms_key.examplekms.arn
          }`,
        ],
        [
          "resource references with lists",
          `
          resource "aws_kms_key" "examplekms" {
            description             = "KMS key 1"
            deletion_window_in_days = 7
          }
          
          resource "aws_s3_bucket" "examplebucket" {
            bucket = "examplebuckettftest"
            acl    = "private"
          }
          
          resource "aws_s3_bucket_object" "examplebucket_object" {
            key        = "someobject"
            bucket     = aws_s3_bucket.examplebucket.*.id
            source     = "index.html"
            kms_key_id = aws_kms_key.examplekms.arn
          }`,
        ],
        [
          "conditionals",
          `
          resource "aws_kms_key" "examplekms" {
            description             = "KMS key 1"
            deletion_window_in_days = 7
          }
          
          resource "aws_s3_bucket" "examplebucket" {
            bucket = "examplebuckettftest"
            acl    = "private"
          }
          
          resource "aws_s3_bucket_object" "examplebucket_object" {
            key        = "someobject"
            bucket     = resource.aws_kms_key.deletion_window_in_days > 3 ? aws_s3_bucket.examplebucket.id : []
            source     = "index.html"
            kms_key_id = aws_kms_key.examplekms.arn
          }`,
        ],
        [
          "for expression 1",
          `
          variable "users" {
            type = map(object({
              is_admin = boolean
            }))
          }
          
          locals {
            admin_users = {
              for name, user in var.users : name => user
              if user.is_admin
            }
            regular_users = {
              for name, user in var.users : name => user
              if !user.is_admin
            }
          }`,
        ],
        [
          "for expression 2",
          `
          variable "users" {
            type = map(object({
              role = string
            }))
          }
          
          locals {
            users_by_role = {
              for name, user in var.users : user.role => name...
            }
          }`,
        ],
        [
          "arithmetics",
          `
          variable "members" {
            type = number
          }
          variable "admins" {
            type = number
          }
          
          locals {
            users = var.members + var.admins
          }`,
        ],
        [
          "for_each loops",
          `
          variable "users" {
            type = set(string)
          }

          resource "aws_iam_user" "lb" {
            for_each = var.users
            name = each.key
            path = "/system/"
          
            tags = {
              tag-key = "tag-value"
            }
          }
          `,
        ],
        [
          "count loops",
          `
          variable "users" {
            type = set(string)
          }

          resource "aws_iam_user" "lb" {
            count = length(var.users)
            name = element(var.users, count.index)
            path = "/system/"
          
            tags = {
              tag-key = "tag-value"
            }
          }`,
        ],
        [
          "dynamic blocks",
          `
          variable "settings" {
            type = list(map(string))
          }

          resource "aws_elastic_beanstalk_environment" "tfenvtest" {
            name                = "tf-test-name"
            application         = "best-app"
            solution_stack_name = "64bit Amazon Linux 2018.03 v2.11.4 running Go 1.12.6"
          
            dynamic "setting" {
              for_each = var.settings
              content {
                namespace = setting.value["namespace"]
                name = setting.value["name"]
                value = setting.value["value"]
              }
            }
          }`,
        ],
      ])("%s", async (_name, hcl) => {
        expect(
          convert(`file.hcl`, hcl, {
            language: "typescript",
          })
        ).rejects.toThrowErrorMatchingSnapshot();
      });
    });
  });
});
