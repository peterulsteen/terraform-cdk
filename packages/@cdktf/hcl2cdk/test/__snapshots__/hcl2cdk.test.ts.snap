// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert typescript complex resource configuration 1`] = `
"new aws.CloudfrontDistribution(this, \\"s3_distribution\\", {
  aliases: [\\"mysite.example.com\\", \\"yoursite.example.com\\"],
  comment: \\"Some comment\\",
  defaultCacheBehavior: [
    {
      allowedMethods: [
        \\"DELETE\\",
        \\"GET\\",
        \\"HEAD\\",
        \\"OPTIONS\\",
        \\"PATCH\\",
        \\"POST\\",
        \\"PUT\\",
      ],
      cachedMethods: [\\"GET\\", \\"HEAD\\"],
      defaultTtl: 3600,
      forwardedValues: [
        {
          cookies: [
            {
              forward: \\"none\\",
            },
          ],
          queryString: false,
        },
      ],
      maxTtl: 86400,
      minTtl: 0,
      targetOriginId: \\"local_s3_origin_id\\",
      viewerProtocolPolicy: \\"allow-all\\",
    },
  ],
  defaultRootObject: \\"index.html\\",
  enabled: true,
  isIpv6Enabled: true,
  loggingConfig: [
    {
      bucket: \\"mylogs.s3.amazonaws.com\\",
      includeCookies: false,
      prefix: \\"myprefix\\",
    },
  ],
  orderedCacheBehavior: [
    {
      allowedMethods: [\\"GET\\", \\"HEAD\\", \\"OPTIONS\\"],
      cachedMethods: [\\"GET\\", \\"HEAD\\", \\"OPTIONS\\"],
      compress: true,
      defaultTtl: 86400,
      forwardedValues: [
        {
          cookies: [
            {
              forward: \\"none\\",
            },
          ],
          headers: [\\"Origin\\"],
          queryString: false,
        },
      ],
      maxTtl: 31536000,
      minTtl: 0,
      pathPattern: \\"/content/immutable/*\\",
      targetOriginId: \\"local_s3_origin_id\\",
      viewerProtocolPolicy: \\"redirect-to-https\\",
    },
    {
      allowedMethods: [\\"GET\\", \\"HEAD\\", \\"OPTIONS\\"],
      cachedMethods: [\\"GET\\", \\"HEAD\\"],
      compress: true,
      defaultTtl: 3600,
      forwardedValues: [
        {
          cookies: [
            {
              forward: \\"none\\",
            },
          ],
          queryString: false,
        },
      ],
      maxTtl: 86400,
      minTtl: 0,
      pathPattern: \\"/content/*\\",
      targetOriginId: \\"local_s3_origin_id\\",
      viewerProtocolPolicy: \\"redirect-to-https\\",
    },
  ],
  origin: [
    {
      domainName: \\"aws_s3_bucket.b.bucket_regional_domain_name\\",
      originId: \\"local_s3_origin_id\\",
      s3OriginConfig: [
        {
          originAccessIdentity:
            \\"origin-access-identity/cloudfront/ABCDEFG1234567\\",
        },
      ],
    },
  ],
  priceClass: \\"PriceClass_200\\",
  restrictions: [
    {
      geoRestriction: [
        {
          locations: [\\"US\\", \\"CA\\", \\"GB\\", \\"DE\\"],
          restrictionType: \\"whitelist\\",
        },
      ],
    },
  ],
  tags: {
    environment: \\"production\\",
  },
  viewerCertificate: [
    {
      cloudfrontDefaultCertificate: true,
    },
  ],
});
"
`;

exports[`convert typescript data references configuration 1`] = `
"const varBucketName = new TerraformVariable(this, \\"bucket_name\\", {
  default: \\"demo\\",
});
const dataAwsS3BucketExamplebucket = new aws.S3Bucket(this, \\"examplebucket\\", {
  acl: \\"private\\",
  bucket: varBucketName,
});
new aws.S3BucketObject(this, \\"examplebucket_object\\", {
  bucket: dataAwsS3Bucket.examplebucket.arn,
  key: \\"someobject\\",
  source: \\"index.html\\",
});
"
`;

exports[`convert typescript double references configuration 1`] = `
"const varBucketName = new TerraformVariable(this, \\"bucket_name\\", {
  default: \\"demo\\",
});
new aws.S3Bucket(this, \\"examplebucket\\", {
  acl: \\"private\\",
  bucket: varBucketName,
  tags: {
    tagKey: varBucketName,
  },
});
"
`;

exports[`convert typescript empty provider configuration 1`] = `
"import * as docker from \\"./.gen/docker\\";
new docker.DockerProvider(this, \\"docker\\", {});
"
`;

exports[`convert typescript errors on arithmetics 1`] = `"Unsupported Terraform feature found: Arithmetics are not yet supported: \${var.members + var.admins}"`;

exports[`convert typescript errors on conditionals 1`] = `"Unsupported Terraform feature found: Conditionals are not yet supported: \${resource.aws_kms_key.deletion_window_in_days > 3 ? aws_s3_bucket.examplebucket.id : []}"`;

exports[`convert typescript errors on count loops 1`] = `"Unsupported Terraform feature found: for-each loops are not yet supported: {\\"count\\":\\"\${length(var.users)}\\",\\"name\\":\\"\${element(var.users, count.index)}\\",\\"path\\":\\"/system/\\",\\"tags\\":{\\"tag-key\\":\\"tag-value\\"}}"`;

exports[`convert typescript errors on dynamic blocks 1`] = `"Unsupported Terraform feature found: for-each loops are not yet supported: {\\"content\\":[{\\"name\\":\\"\${setting.value[\\\\\\"name\\\\\\"]}\\",\\"namespace\\":\\"\${setting.value[\\\\\\"namespace\\\\\\"]}\\",\\"value\\":\\"\${setting.value[\\\\\\"value\\\\\\"]}\\"}],\\"for_each\\":\\"\${var.settings}\\"}"`;

exports[`convert typescript errors on for expression 1 1`] = `
"Unsupported Terraform feature found: For expressions are not yet supported: \${{
              for name, user in var.users : name => user
              if user.is_admin
            }}"
`;

exports[`convert typescript errors on for expression 2 1`] = `
"Unsupported Terraform feature found: For expressions are not yet supported: \${{
              for name, user in var.users : user.role => name...
            }}"
`;

exports[`convert typescript errors on for_each loops 1`] = `"Unsupported Terraform feature found: for-each loops are not yet supported: {\\"for_each\\":\\"\${var.users}\\",\\"name\\":\\"\${each.key}\\",\\"path\\":\\"/system/\\",\\"tags\\":{\\"tag-key\\":\\"tag-value\\"}}"`;

exports[`convert typescript errors on resource references with HCL functions 1`] = `"Unsupported Terraform feature found: Functions are not yet supported: \${element(aws_s3_bucket.examplebucket, 0).id}"`;

exports[`convert typescript errors on resource references with lists 1`] = `"Unsupported Terraform feature found: Splat operations (resource.name.*.property) are not yet supported: \${aws_s3_bucket.examplebucket.*.id}"`;

exports[`convert typescript locals configuration 1`] = `
"const howMany = 42;
const isItGreat = true;
const owner = \\"Community Team\\";
const serviceName = \\"forum\\";
"
`;

exports[`convert typescript locals references configuration 1`] = `
"const bucketName = \\"foo\\";
new aws.S3Bucket(this, \\"examplebucket\\", {
  acl: \\"private\\",
  bucket: localBucketName,
});
"
`;

exports[`convert typescript modules configuration 1`] = `
"import * as TerraformAwsModulesVpcAws from \\"./.gen/terraform-aws-modules/vpc/aws\\";
new TerraformAwsModulesVpcAws(this, \\"vpc\\", {
  azs: [\\"eu-west-1a\\", \\"eu-west-1b\\", \\"eu-west-1c\\"],
  cidr: \\"10.0.0.0/16\\",
  enableNatGateway: true,
  enableVpnGateway: true,
  name: \\"my-vpc\\",
  privateSubnets: [\\"10.0.1.0/24\\", \\"10.0.2.0/24\\", \\"10.0.3.0/24\\"],
  publicSubnets: [\\"10.0.101.0/24\\", \\"10.0.102.0/24\\", \\"10.0.103.0/24\\"],
  tags: {
    environment: \\"dev\\",
    terraform: \\"true\\",
  },
});
"
`;

exports[`convert typescript multiple locals blocks configuration 1`] = `
"const owner = \\"Community Team\\";
const serviceName = \\"forum\\";
const howMany = 42;
const isItGreat = true;
"
`;

exports[`convert typescript multiple outputs configuration 1`] = `
"new TerraformOutput(this, \\"first_cidr_out\\", {
  value: \\"first\\",
});
new TerraformOutput(this, \\"second_cidr_out\\", {
  value: \\"second\\",
});
"
`;

exports[`convert typescript output configuration 1`] = `
"new TerraformOutput(this, \\"cidr_out\\", {
  value: \\"test\\",
});
"
`;

exports[`convert typescript output withdescription configuration 1`] = `
"new TerraformOutput(this, \\"cidr_out\\", {
  value: \\"test\\",
  description: \\"Best output\\",
  sensitive: true,
});
"
`;

exports[`convert typescript provider with complex config configuration 1`] = `
"import * as aws from \\"./.gen/aws\\";
new aws.AwsProvider(this, \\"aws\\", {
  accessKey: \\"mock_access_key\\",
  endpoints: [
    {
      dynamodb: \\"http://localhost:8000\\",
    },
  ],
  region: \\"us-east-1\\",
  secretKey: \\"mock_secret_key\\",
  skipCredentialsValidation: true,
  skipMetadataApiCheck: true,
  skipRequestingAccountId: true,
});
"
`;

exports[`convert typescript referenced modules configuration 1`] = `
"import * as TerraformAwsModulesVpcAws from \\"./.gen/terraform-aws-modules/vpc/aws\\";
const moduleVpc = new TerraformAwsModulesVpcAws(this, \\"vpc\\", {
  azs: [\\"eu-west-1a\\", \\"eu-west-1b\\", \\"eu-west-1c\\"],
  cidr: \\"10.0.0.0/16\\",
  enableNatGateway: true,
  enableVpnGateway: true,
  name: \\"my-vpc\\",
  privateSubnets: [\\"10.0.1.0/24\\", \\"10.0.2.0/24\\", \\"10.0.3.0/24\\"],
  publicSubnets: [\\"10.0.101.0/24\\", \\"10.0.102.0/24\\", \\"10.0.103.0/24\\"],
  tags: {
    environment: \\"dev\\",
    terraform: \\"true\\",
  },
});
new TerraformOutput(this, \\"subnet_ids\\", {
  value: moduleVpc.publicSubnetsOutput,
});
"
`;

exports[`convert typescript resource references configuration 1`] = `
"const awsKmsKeyExamplekms = new aws.KmsKey(this, \\"examplekms\\", {
  deletionWindowInDays: 7,
  description: \\"KMS key 1\\",
});
const awsS3BucketExamplebucket = new aws.S3Bucket(this, \\"examplebucket\\", {
  acl: \\"private\\",
  bucket: \\"examplebuckettftest\\",
});
new aws.S3BucketObject(this, \\"examplebucket_object\\", {
  bucket: awsS3BucketExamplebucket.id,
  key: \\"someobject\\",
  kmsKeyId: awsKmsKeyExamplekms.arn,
  source: \\"index.html\\",
});
"
`;

exports[`convert typescript sensitive output configuration 1`] = `
"new TerraformOutput(this, \\"cidr_out\\", {
  value: \\"test\\",
  sensitive: true,
});
"
`;

exports[`convert typescript simple data source configuration 1`] = `
"new aws.Subnet(this, \\"selected\\", {
  id: \\"subnet_id\\",
});
"
`;

exports[`convert typescript simple resource configuration 1`] = `
"new aws.Vpc(this, \\"example\\", {
  cidrBlock: \\"10.0.0.0/16\\",
});
"
`;

exports[`convert typescript variable configuration 1`] = `
"new TerraformVariable(this, \\"weekday\\", {});
"
`;

exports[`convert typescript variable references configuration 1`] = `
"const varBucketName = new TerraformVariable(this, \\"bucket_name\\", {
  default: \\"demo\\",
});
new aws.S3Bucket(this, \\"examplebucket\\", {
  acl: \\"private\\",
  bucket: varBucketName,
});
"
`;

exports[`convert typescript variable with default configuration 1`] = `
"new TerraformVariable(this, \\"availability_zone_names\\", {
  default: [\\"us-west-1a\\"],
  description: \\"What AMI to use to create an instance\\",
});
"
`;
